lib/cea/cea608_memory.js

```diff
@@ -0,0 +1,513 @@
+   */
+  addChar(set, b) {
+    // Valid chars are in the range [0x20, 0x7f]
+    if (b < 0x20 || b > 0x7f) {
+      return;
+    }
+
+    let char = '';
+    switch (set) {
+      case shaka.cea.Cea608Memory.CharSet.BASIC_NORTH_AMERICAN:
+        if (shaka.cea.Cea608Memory.CharSet.BasicNorthAmericanChars.has(b)) {
+          char =
+                shaka.cea.Cea608Memory.CharSet.BasicNorthAmericanChars.get(b);
+        } else {
+          // Regular ASCII
+          char = String.fromCharCode(b);
+        }
+        break;
+      case shaka.cea.Cea608Memory.CharSet.SPECIAL_NORTH_AMERICAN:
+        char =
+              shaka.cea.Cea608Memory.CharSet.SpecialNorthAmericanChars.get(b);
+        break;
+      case shaka.cea.Cea608Memory.CharSet.SPANISH_FRENCH:
+        // Extended charset does a BS over preceding char, 6.4.2 EIA-608-B.
+        this.eraseChar();
+        char =
+              shaka.cea.Cea608Memory.CharSet.ExtendedSpanishFrench.get(b);
+        break;
+      case shaka.cea.Cea608Memory.CharSet.PORTUGUESE_GERMAN:
+        this.eraseChar();
+        char =
+              shaka.cea.Cea608Memory.CharSet.ExtendedPortugueseGerman.get(b);
+        break;
+    }
+
+    if (char) {
+      const styledChar = new shaka.cea.Cea608Char(char, this.underline_,
+          this.italics_, this.backgroundColor_, this.textColor_);
+      this.rows_[this.row_].push(styledChar);
+    }
+  }
+
+  /**
+   * Erases a character from the buffer.
+   */
+  eraseChar() {
+    this.rows_[this.row_].pop();
+  }
+
+  /**
+   * Moves rows of characters.
+   * @param {!number} dst Destination row index.
+   * @param {!number} src Source row index.
+   * @param {!number} count Count of rows to move.
+   */
+  moveRows(dst, src, count) {
+    for (let i = 0; i < count; i++) {
+      this.rows_[dst + i] = this.rows_[src + i];
+    }
+  }
+
+  /**
+   * Resets rows of characters.
+   * @param {!number} idx Starting index.
+   * @param {!number} count Count of rows to reset.
+   */
+  resetRows(idx, count) {
+    for (let i = 0; i <= count; i++) {
+      this.rows_[idx + i] = [];
+    }
+  }
+
+  /**
+   * Resets the entire memory buffer.
+   */
+  resetAllRows() {
+    this.resetRows(0, shaka.cea.Cea608Memory.CC_ROWS);
+  }
+
+  /**
+   * Erases entire memory buffer.
+   * Doesn't change scroll state or number of rows.
+   */
+  eraseBuffer() {
+    this.row_ = (this.scrollRows_ > 0) ? this.scrollRows_ : 1;
+    this.resetAllRows();
+  }
+
+  /**
+   * @param {!boolean} underline
+   */
+  setUnderline(underline) {
+    this.underline_ = underline;
+  }
+
+  /**
+   * @param {!boolean} italics
+   */
+  setItalics(italics) {
+    this.italics_ = italics;
+  }
+
+  /**
+   * @param {!string} color
+   */
+  setTextColor(color) {
+    this.textColor_ = color;
+  }
+
+  /**
+   * @param {!string} color
+   */
+  setBackgroundColor(color) {
+    this.backgroundColor_ = color;
+  }
+
+  /**
+   * @return {!shaka.cea.CeaDecoder}
+   */
+  getDecoder() {
```

Comment:
Could you do without this to try to break the circular dependencies?  At the moment I'm writing this comment, I've seen getDecoder() used in tests to extract output, but I haven't read through the CeaDecoder or Cea608DataChannel yet.  So I don't yet know if there is some other need for it.